### ISSUE TITLE
MiniMagick::Image @tempfile exists after GC?

### ISSUE THREAD

**Issue by fuminori-ido-m** [minimagick/minimagick#555]
MiniMagick::Image.create creates @tempfile (ruby Tempfile object) at:
> tempfile = MiniMagick::Utilities.tempfile(ext.to_s.downcase, &block)

at MiniMagick::Utilities#tempfile:
> Tempfile.new(["mini_magick", ...

Is this @tempfile guaranteed to exist during minimagick operation (e.g. convert)?  I'm asking this question because:

1. This temporary file is sometimes disappeared under high CPU pressure in our application.
2. ruby Tempfile is deleted at GC and at exit according to [ruby doc](https://ruby-doc.org/stdlib-2.5.6/libdoc/tempfile/rdoc/Tempfile.html).

**Comment by janko** [comment]
The tempfile should only deleted by GC if nothing is referencing the `Tempfile` object anymore. MiniMagick is internally implemented to keep a reference while commands are being executed. Is there any specific code you're concerned about?

**Comment by fuminori-ido-m** [comment]
Thank you for your prompt reply, @janko !

> MiniMagick is internally implemented to keep a reference while commands are being executed

I think so too and I couldn't reproduce tempfile disappearance while enforce GC.start in my debug console during minimagick operation.  But... I saw sometimes tempfile  disappearance happened during minimagick convert operation under CPU high pressure.

It could be another reason of tempfile disappearance (maybe disk size shortage(even though I already checked it), or something??) so please close this issue if you do not see any problem.

**Comment by whunter** [comment]
Hi @janko and @fuminori-ido-m 

Is this an example of the tempfile that you're discussing? 
```/usr/local/bundle/gems/mini_magick-5.0.0/lib/mini_magick/shell.rb:20:in `run': `identify -format %m %w %h %b /tmp/mini_magick20240711-15-gy4kf3.jpg[0]` failed with status: 1 and error: (MiniMagick::Error)

identify-im6.q16: unable to open image `/tmp/mini_magick20240711-15-gy4kf3.jpg': No such file or directory @ error/blob.c/OpenBlob/2874.```

We're using a no longer maintained [gem](https://github.com/cmoa/iiif_s3) to generate [iiif](https://iiif.io/) compatible tiles and it uses `"mini_magick", ">= 4.8"` so I guess there's an entire layer between what I'm seeing and what you wrote maybe a decade ago? Admittedly, not ideal

We started seeing this error when we refactored from adding all images to a data object and passing them into that gem to be processed, to iteratively downloading one image, processing it, and cleaning up. I don't know the details yet on how that gem calls MiniMagick and I don't expect you to either. But that seems like the kind of refactor that could lead to more frequent GC to me. That was actually one of our motivations. 

So basically, I don't expect you all to debug this or anything, but I figured I'd at least ask if you're just familiar w/ this error and if the GC thing could be the cause.

Thanks for your time.

**Comment by fuminori-ido** [comment]
I found that [Ruby Tempfile document (in Japanese) example showed to disable GC](https://docs.ruby-lang.org/ja/latest/method/Tempfile/s/create.html) as follows:
```ruby
GC.disable
path = ""
Tempfile.create("foo") do |f|
  path = f.path
  p File.exist?(path) #=> true
end
```
Unfortunattely I couldn't find out the same example in English version of ruby document, but the example above means that GC is required to stop around Minimagick process also.

**Comment by janko** [comment]
MiniMagick doesn't use `Tempfile.create`, it uses `Tempfile.new`. AFAICT `Tempfile.create` doesn't create a `Tempfile` instance, which means there are no GC implications on its behavior (only `Tempfile#initialize` defines GC finalizers).

**Comment by whunter** [comment]
Thanks for the responses. Very much appreciated. I'll keep digging
